app/uk/gov/hmrc/hec/controllers/IFController.scala

```diff
@@ -0,0 +1,130 @@
+import uk.gov.hmrc.hec.services.IFService
+import uk.gov.hmrc.hec.util.Logging
+import uk.gov.hmrc.hec.util.Logging.LoggerOps
+import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class IFController @Inject() (
+  IFService: IFService,
+  cc: ControllerComponents
+)(implicit ec: ExecutionContext)
+    extends BackendController(cc)
+    with Logging {
+
+  private def getError[A](optional: Option[A], error: String) = if (optional.isEmpty) Some(error) else None
+  private def collateErrors(errors: Seq[Option[String]])      = errors.collect({ case Some(v) => v }).mkString("; ")
+
+  /**
+    * Fetch individual user's self-assessment status
+    * @param utr Self-assessment UTR
+    * @param taxYear Start tax year
+    * @return Self-assessment status with UTR and tax year
+    */
+  def getSAStatus(utr: String, taxYear: String): Action[AnyContent] = Action.async { implicit request =>
+    val validSAUTROpt   = SAUTR.fromString(utr)
+    val validTaxYearOpt = TaxYear.fromString(taxYear)
+
+    (validSAUTROpt, validTaxYearOpt) match {
+      case (Some(utr), Some(year)) =>
+        IFService
+          .getSAStatus(utr, year)
+          .fold(
+            { e =>
+              logger.warn("Could not fetch SA status", e)
+              InternalServerError
+            },
+            saStatus => Ok(Json.toJson(saStatus))
+          )
+
+      case (utr, year) =>
+        Future.successful(
+          BadRequest(
+            collateErrors(
+              Seq(
+                getError(utr, "Invalid SAUTR"),
+                getError(year, "Invalid tax year")
+              )
+            )
+          )
+        )
+    }
+  }
+
+  /**
+    * Fetch a company's corporation tax status
+    * @param utr The corporation tax UTR
+    * @param fromDate The start date of the tax period
+    * @param toDate The end date of the tax period
+    * @return Corporation tax status & accounting periods
+    */
+  def getCTStatus(
+    utr: String,
+    fromDate: String,
+    toDate: String
+  ): Action[AnyContent] = Action.async { implicit request =>
+    val validCTUTROpt = CTUTR.fromString(utr)
+
+    def parsedDate(dateStr: String): Option[LocalDate] = try Some(LocalDate.parse(dateStr, DateTimeFormatter.ISO_DATE))
+    catch {
+      case _: Exception => None
+    }
+
+    val validFromDateOpt = parsedDate(fromDate)
+    val validToDateOpt   = parsedDate(toDate)
+
+    (validCTUTROpt, validFromDateOpt, validToDateOpt) match {
+      case (Some(utr), Some(from), Some(to)) =>
+        IFService
+          .getCTStatus(utr, from, to)
+          .fold(
+            { e =>
+              logger.warn("Could not fetch CT status", e)
+              InternalServerError
+            },
+            ctStatus => Ok(Json.toJson(ctStatus))
+          )
+
+      case (utr, from, to) =>
+        Future.successful(
+          BadRequest(
+            collateErrors(
+              Seq(
```

Comment:
you could use the `ValidatedNel` pattern with `mapN` to collect validation errors more neatly. Basically you'd end up with something like:
```scala
import cats.syntax.apply._

val ctutrValidation: ValidatedNel[String, CTUTR] = ???
val fromDateValidation: ValidatedNel[String, LocalDate] = ???
val toDateValidation: ValidatedNel[String, LocalDate] = ???

val validation = (ctutrValidation, fromDateValidation, toDateValidation).mapN((_, _, _))
validation match {
  case Valid(...) =>
  case Invalid(e) => BadRequest(e.toList.mkString(";")
}
```

Reply:
yeah, I saw this in @priyankadhiman3006's PR, is this something we do across the code base, I was just trying to avoid importing `cats`. I'm guessing it's not too big a library so should be ok use?

Reply:
it's a useful feature I think that we can make use of - `cats` is already being included in the dependencies so it should be ready for you to pick up. There's things like `EitherT` which we're already making use of

Reply:
of course, don't know why I thought I had to add the dependency, ignore me!

Reply:
updated in commit - https://github.com/hmrc/hec/pull/3/commits/8d7b8e4edd250abbf311dd9a2a789f3c189d708d